### ISSUE TITLE
Compiling an infinite recursion call to main function in main function by modules

### ISSUE THREAD

**Issue by vickyqt** [rust-lang/rust#151434]
I tried this code:

```rust
mod restmd {
    pub fn rst() {
        print!("  Heey  ");
        super::main();
    }
}

fn main() {
    restmd::rst();
}

```

I expected to see an error while compilation for an infinite recursion call.

Instead, it compiled and I got main function stack overflowed.

### Meta

`rustc --version --verbose`:
```

<!-- Failed to upload "rust_stoflo.mp4" -->

rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-unknown-linux-gnu
release: 1.92.0
LLVM version: 21.1.3
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by hanna-kruppe** [comment]
I don’t think that there’s anything special about the main function or modules in this respect. The unconditional recursion lint only catches functions calling themselves directly, any recursion through two or more functions will be missed. There is an existing issue for improving this https://github.com/rust-lang/rust/issues/57965
